backend/src/main/java/pl/zzpj/esportbetting/impl/UserLevelServiceImpl.java

```diff
@@ -0,0 +1,65 @@
+package pl.zzpj.esportbetting.impl;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import pl.zzpj.esportbetting.interfaces.UserLevelService;
+import pl.zzpj.esportbetting.model.User;
+import pl.zzpj.esportbetting.repos.LevelRepository;
+import pl.zzpj.esportbetting.repos.UserRepository;
+import pl.zzpj.esportbetting.strategies.ExpGiverStrategy;
+import pl.zzpj.esportbetting.strategies.HappyHoursStrategy;
+import pl.zzpj.esportbetting.strategies.NormalStrategy;
+
+import java.time.LocalTime;
+
+@Service("userLevelService")
+public class UserLevelServiceImpl implements UserLevelService {
+
+    private static final int EXP_AFTER_WINNING_BET = 200;
+    private static final int EXP_AFTER_LOOSING_BET = 50;
+
+    @Value("#{T(java.time.LocalTime).parse('${esportbetting.app.happy-hours.from}')}")
+    private LocalTime happyHoursFrom;
+
+    @Value("#{T(java.time.LocalTime).parse('${esportbetting.app.happy-hours.to}')}")
+    private LocalTime happyHoursTo;
+
+    private final LevelRepository levelRepository;
+    private final UserRepository userRepository;
+    private ExpGiverStrategy expGiverStrategy;
+
+    @Autowired
+    UserLevelServiceImpl(LevelRepository levelRepository, UserRepository userRepository) {
+        this.levelRepository = levelRepository;
+        this.userRepository = userRepository;
+    }
+
+    @Override
+    public void addExpAfterBet(User user, boolean won) {
+        int expToAdd = 0;
+        if (won) {
+            expToAdd = EXP_AFTER_WINNING_BET;
+        } else {
+            expToAdd = EXP_AFTER_LOOSING_BET;
+        }
+        if (checkIfHappyHourNow()) {
+            user.addExp(new HappyHoursStrategy().calcExp(expToAdd));
+        } else {
+            user.addExp(new NormalStrategy().calcExp(expToAdd));
+        }
```

Comment:
To nie jest poprawna implementacja wzorca strategi.
Zmienna ```expFiverStrategy``` w ogóle nie jest używana, a konkretne strategie są tworzone wewnątrz metody, co całkowicie uniemożliwia nam niezależne rozszerzenie strategii o nowe klasy, a właśnie w to chcemy osiągnąć w tym wzorcu. 
W tym wzorcu  w klasie, której używasz strategii (tutaj to będzie ```BetServiceImpl``` w metodzie ```withdrawBetMoneyAndExpForMatch()```)  musisz w ustawić metodą ```set()``` obiekt strategii.
Wtedy dodanie expa będzie wyglądało tak:
```user.addExp(this.expGiverStrategy.calcExp(expToAdd))```

Reply:
okej dodam setter i strategy jako pole klasy